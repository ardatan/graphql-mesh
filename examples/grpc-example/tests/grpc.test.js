const { findAndParseConfig } = require('@graphql-mesh/config');
const { getMesh } = require('@graphql-mesh/runtime');
const { basename, join } = require('path');

const { introspectionFromSchema, lexicographicSortSchema } = require('graphql');
const { readFile } = require('fs-extra');

const config$ = findAndParseConfig({
  dir: join(__dirname, '..'),
});
const mesh$ = config$.then(config => getMesh(config));
const startGrpcServer = require('../start-server');
const grpc$ = startGrpcServer(300);
jest.setTimeout(15000);

describe('gRPC Example', () => {
  it('should generate correct schema', async () => {
    const { schema } = await mesh$;
    expect(
      introspectionFromSchema(lexicographicSortSchema(schema), {
        descriptions: false,
      })
    ).toMatchSnapshot('grpc-schema');
  });
  it('should get movies correctly', async () => {
    const GetMoviesQuery = await readFile(join(__dirname, '../example-queries/GetMovies.query.graphql'), 'utf8');
    const { execute } = await mesh$;
    await grpc$;
    const result = await execute(GetMoviesQuery);
    expect(result).toMatchSnapshot('get-movies-grpc-example-result');
  });
  it('should fetch movies by cast as a subscription correctly', async () => {
    const MoviesByCastSubscription = await readFile(join(__dirname, '../example-queries/MoviesByCast.subscription.graphql'), 'utf8');
    const { subscribe } = await mesh$;
    await grpc$;
    const resultIterator = await subscribe(MoviesByCastSubscription);
    expect(Symbol.asyncIterator in resultIterator).toBeTruthy();
    expect(await resultIterator.next()).toMatchSnapshot('movies-by-cast-grpc-example-result-1');
    expect(await resultIterator.next()).toMatchSnapshot('movies-by-cast-grpc-example-result-2');
    await resultIterator.return();
  })
  afterAll(() => {
      mesh$.then(mesh => mesh.destroy());
      grpc$.then(grpc => grpc.forceShutdown());
  });
});
