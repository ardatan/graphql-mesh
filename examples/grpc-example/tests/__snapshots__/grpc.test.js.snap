// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRPC Example should generate correct schema: grpc-schema 1`] = `
"\\"\\"\\"
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
\\"\\"\\"
scalar BigInt

enum ConnectivityState {
  CONNECTING
  IDLE
  READY
  SHUTDOWN
  TRANSIENT_FAILURE
}

type Mutation {
  \\"\\"\\"get all movies\\"\\"\\"
  ioXtechAnotherExampleGetMovies(input: io_xtech_MovieRequest_Input): io_xtech_MoviesResult

  \\"\\"\\"get all movies\\"\\"\\"
  ioXtechExampleGetMovies(input: io_xtech_MovieRequest_Input): io_xtech_MoviesResult
}

type Query {
  ioXtechAnotherExampleConnectivityState(tryToConnect: Boolean): ConnectivityState
  ioXtechExampleConnectivityState(tryToConnect: Boolean): ConnectivityState
}

type Subscription {
  \\"\\"\\"search movies by the name of the cast\\"\\"\\"
  ioXtechAnotherExampleSearchMoviesByCast(input: io_xtech_SearchByCastRequest_Input): io_xtech_Movie

  \\"\\"\\"search movies by the name of the cast\\"\\"\\"
  ioXtechExampleSearchMoviesByCast(input: io_xtech_SearchByCastRequest_Input): io_xtech_Movie
}

type google_protobuf_Timestamp {
  nanos: Int
  seconds: BigInt
}

input google_protobuf_Timestamp_Input {
  nanos: Int
  seconds: BigInt
}

enum io_xtech_Genre {
  ACTION
  DRAMA
  UNSPECIFIED
}

\\"\\"\\"movie message payload\\"\\"\\"
type io_xtech_Movie {
  \\"\\"\\"list of cast\\"\\"\\"
  cast: [String]
  genre: io_xtech_Genre
  name: String
  rating: Float
  time: google_protobuf_Timestamp
  year: BigInt
}

input io_xtech_MovieRequest_Input {
  movie: io_xtech_Movie_Input
}

\\"\\"\\"movie message payload\\"\\"\\"
input io_xtech_Movie_Input {
  \\"\\"\\"list of cast\\"\\"\\"
  cast: [String]
  genre: io_xtech_Genre
  name: String
  rating: Float
  time: google_protobuf_Timestamp_Input
  year: BigInt
}

\\"\\"\\"movie result message, contains list of movies\\"\\"\\"
type io_xtech_MoviesResult {
  \\"\\"\\"list of movies\\"\\"\\"
  result: [io_xtech_Movie]
}

input io_xtech_SearchByCastRequest_Input {
  castName: String
}
"
`;

exports[`gRPC Example should get movies correctly: get-movies-grpc-example-result 1`] = `
Object {
  "data": Object {
    "ioXtechExampleGetMovies": Object {
      "result": Array [
        Object {
          "cast": Array [
            "Leonardo DiCaprio",
            "Jonah Hill",
            "Margot Robbie",
          ],
          "genre": "DRAMA",
          "name": "The Wolf of Wall Street",
          "rating": 0.7799999713897705,
          "time": Object {
            "seconds": 1608422400000n,
          },
          "year": 2013n,
        },
      ],
    },
  },
}
`;
